source/common/tcp_proxy/tcp_proxy.h

```diff
@@ -117,7 +118,16 @@ class TunnelingConfigHelperImpl : public TunnelingConfigHelper {
           config_message)
       : hostname_(config_message.hostname()), use_post_(config_message.use_post()),
         header_parser_(Envoy::Router::HeaderParser::configure(config_message.headers_to_add())) {}
-  const std::string& hostname() const override { return hostname_; }
+  const std::string host(const StreamInfoOptPtr downstream_info = absl::nullopt) const override {
+    size_t colon_idx = hostname_.find(':');
+    std::string hostname = hostname_.substr(0, colon_idx);
+    int port = std::stoi(hostname_.substr(colon_idx + 1));
+    if (downstream_info.has_value() && hostname == "%REQUESTED_SERVER_NAME%") {
+      return fmt::sprintf(
+          "%s:%d", downstream_info->get().downstreamAddressProvider().requestedServerName(), port);
```

Comment:
I wonder why we don't just support the entire formatter dictionary here? It's a bit strange and error-prone to just pick one variable.

Reply:
yeah it's great to see this hanging together end to end, but the implementation seems pretty fragile.  I'd suggest a standalone utility to swap the host part of the host-part pair, 
Also maybe a string_view for requested_server_name and a utility to handle the optional deref rather than passing all the downstream info through, WDYT?
/wait

Reply:
I don't see a need to support values like: `%START_TIME%` or `%UPSTREAM_PROTOCOL%`. Why would anyone need other values from dictionary?

Reply:
@jewertow Yes, those are not applicable, but I think headers, things like metadata, filter state, or other TLS attributes (DNS SAN) are reasonable.

Reply:
Unfortunately headers couldn't be used, because it's a TcpProxy, not an HttpConnectionManager, so these data are not available. When it comes to TLS attributes, there are only available SANs of downstream proxy (`%DOWNSTREAM_LOCAL_URI_SAN%`, `%DOWNSTREAM_PEER_URI_SAN%`) and using these values does not make sense, because it would connect back to the downstream - I can't imagine any real use case. Other TLS attributes don't make sense in the context of hostname.

I agree that dynamic metadata or filter state might somehow be useful to provide hostname, but only these two. I considered all other command operators and can't imagine that any of them would be useful in the context of hostname. So I don't agree that supporting all command operators is important, because almost all of them are useless and it's not worth to implement it, test all cases and then maintain useless features and complex code base.

> It's a bit strange and error-prone to just pick one variable.

I don't agree that it's error-prone, because passing a command operator which is evaluated to irrelevant value is also strange error-prone. I'm going to document it, so it won't be confusing nor strange.

I understand your motivation and respect your feedback, but I just don't believe in this idea and I really think it's useless.
Dynamic metadata and filter state might be somehow useful, but I suggest to not support them as far as users don't need them. This PR aims to solve a real and common problem, so I am not convinced to extend its scope if it's not needed.

Reply:
I am not disagreeing about the value or that it only makes sense with a subset of formatter strings. I am trying to avoid having special custom parsing code since it's difficult to maintain. I'm OK just supporting SNI for now, and tackle metadata/filter state later. As I mentioned earlier, we'll have to support URL templates for MASQUE which need generic templates.